fairscale/optim/feed_forward2.py

```diff
@@ -0,0 +1,204 @@
+        self.relu = nn.ReLU()
+        self.fc2 = nn.Linear(self.hidden_size, 1)
+        self.sigmoid = nn.Sigmoid()
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:  # type: ignore
+        out = self.fc1(x)
+        out = self.relu(out)
+        out = self.fc2(out)
+        out = self.sigmoid(out)
+        return out
+
+
+class LayerwiseGradientScaler:
+    def __init__(self, modules: nn.Module):
+        self.per_layer_scale = [2 ** (1 / 1000) for _ in modules]
+        # print('length of modules %s' % len(modules))
+
+    def scale_up(self, m: nn.Module, inputs: torch.Tensor, outputs: torch.Tensor) -> None:
+        if inputs[0] is not None:
+            torch.mul(inputs[0], (2 ** 13))
+
+    def scale_down(self, m: nn.Module, inputs: torch.Tensor, outputs: torch.Tensor) -> None:
+        if outputs[0] is not None:
+            torch.mul(outputs[0], 1 / (2 ** 13))
+
+
+# assign labels
+def blob_label(y: np.ndarray, label: int, loc: list) -> np.ndarray:
+    target = np.copy(y)  # type: ignore
+    for l in loc:
+        target[y == l] = label
+    return target
+
+
+def load_data() -> Tuple:
+    torch.manual_seed(11)
+    x_train, y_train = make_blobs(n_samples=40, n_features=2, cluster_std=1.5, shuffle=True, random_state=10)
+    x_train = torch.FloatTensor(x_train)
+    y_train = torch.FloatTensor(blob_label(y_train, 0, [0]))
+    y_train = torch.FloatTensor(blob_label(y_train, 1, [1, 2, 3]))
+
+    x_test, y_test = make_blobs(n_samples=10, n_features=2, cluster_std=1.5, shuffle=True, random_state=10)
+    x_test = torch.FloatTensor(x_test)
+    y_test = torch.FloatTensor(blob_label(y_test, 0, [0]))
+    y_test = torch.FloatTensor(blob_label(y_test, 1, [1, 2, 3]))
+
+    all_data = (x_train, y_train, x_test, y_test)
+    return all_data
+
+
+def standard_training(model: Feedforward) -> Feedforward:
+    criterion = torch.nn.BCELoss()
+    optimizer = torch.optim.SGD(model.parameters(), lr=0.001)
+
+    x_train, y_train, x_test, y_test = load_data()
+    model.train()
+    num_epochs = 2
+    for _ in range(num_epochs):
+        optimizer.zero_grad()
+
+        # forward pass
+        y_pred = model(x_train)
+        # print(y_pred)
+
+        # compute loss
+        loss = criterion(y_pred.squeeze(), y_train)
+
+        # backward pass
+        loss.backward()
+
+        # update weights
+        optimizer.step()
+    return model
+
+
+def test_weights() -> None:
+    model1 = Feedforward(2, 10)
+    model2 = Feedforward(2, 10)
+
+    layers_to_scale = list()
+    for layer in model2.named_children():
+        print(layer)
+        if isinstance(layer[1], nn.Linear):
+            layers_to_scale.append(layer)
+
+    layerwise_grad_scaler = LayerwiseGradientScaler(layers_to_scale)
+    for _, layer in enumerate(layers_to_scale):
+        layer[1].register_full_backward_hook(layerwise_grad_scaler.scale_up)
+        layer[1].register_full_backward_hook(layerwise_grad_scaler.scale_down)
```

Comment:
also why do we need two hooks for these? why not just do scale_up/scale_down in a single hook